src/fsharp/Optimizer.fs

```diff
@@ -1615,6 +1620,78 @@ let rec RearrangeTupleBindings expr fin =
         | None -> None
     | _ -> None
 
+// Attempts to rewrite tuple bindings containing ifs/matches by introducing a mutable local for each tuple element.
+// These are assigned to exactly once from each branch in order to eliminate tuple allocations. The tuple binding
+// is also rearranged such that OptimizeTupleFieldGet may kick in (see RearrangeTupleBindings comment above).
+// First class use of a tuple at the end of any branch prevents this rewrite.
+//
+// Roughly speaking, the following expression:
+//
+//    let a, b =
+//        if cond () then
+//            1, 2
+//        elif cond2 () then
+//            3, 4
+//        else
+//            5, 6
+//    in ...
+//
+// becomes
+//
+//    let mutable a = Unchecked.defaultof<_>
+//    let mutable b = Unchecked.defaultof<_>
+//
+//    if cond () then
+//        a <- 1
+//        b <- 2
+//    elif cond2 () then
+//        a <- 3
+//        b <- 4
+//    else
+//        a <- 5
+//        b <- 6
+//    in ...
+let TryRewriteBranchingTupleBinding g (v: Val) rhs tgtSeqPtOpt body m =
+    let rec dive g m vrefs expr =
+        match expr with
+        | Expr.Match (sp, inputRange, decision, targets, fullRange, ty) ->
+            // Recurse down every if/match branch
+            let rewrittenTargets = targets |> Array.choose (fun (TTarget (vals, targetExpr, sp)) ->
+                match dive g m vrefs targetExpr with
+                | Some rewritten -> TTarget (vals, rewritten, sp) |> Some
+                | _ -> None)
+
+            // If not all branches can be rewritten, keep the original expression as it is
+            if rewrittenTargets.Length <> targets.Length then
+                None
+            else
+                Expr.Match (sp, inputRange, decision, rewrittenTargets, fullRange, ty) |> Some
+        | Expr.Op (TOp.Tuple tupInfo, _, tupleElements, m) when not (evalTupInfoIsStruct tupInfo) ->
+            // Replace tuple allocation with mutations of locals
+            List.map2 (mkValSet m) vrefs tupleElements
+            |> mkSequentials DebugPointAtSequential.StmtOnly g m
+            |> Some
+        | Expr.Sequential (e1, e2, kind, sp, m) ->
+            match dive g m vrefs e2 with
+            | Some rewritten -> Expr.Sequential (e1, rewritten, kind, sp, m) |> Some
+            | _ -> None
+        | Expr.Let (bind, body, m, _) ->
+            match dive g m vrefs body with
+            | Some rewritten -> mkLetBind m bind rewritten |> Some
+            | _ -> None
+        | _ -> None
+
+    let argTys = destRefTupleTy g v.Type
```

Comment:
This has the potential to execute very often - once for each tuple-typed `let`.  I guess the tuple type acts as a good filter but I think it would be great to delay this execution of computing `vrefs` until we know it's really going to be necessary, i.e. we actually have a `Expr.Op` at the end of every branch, or at least one branch.

Reply:
Would basically wrapping these 4 lines in a `Lazy` be an acceptable solution?

Reply:
@dsyme, it's ready for another look.